### ISSUE TITLE
Maximum call stack size exceeded with self-referencing objects on .resolve

### ISSUE THREAD

**Issue by bbqsrc** [APIDevTools/json-schema-ref-parser#36]
The following causes `RangeError: Maximum call stack size exceeded`:

```javascript
const obj = {
  definitions: {
    Self: {
      properties: {
        self: { $ref: "#/definitions/Self" }
      }
    }
  }
}

require("json-schema-ref-parser").resolve(obj, (err, schema) => console.log(err, schema))
```

**Comment by philsturgeon** [comment]
Your JS now outputs:


```
null <ref *1> $Refs {
  circular: false,
  '_$refs': {
    '/private/tmp/': $Ref {
      path: '/private/tmp/',
      value: [Object],
      '$refs': [Circular *1],
      pathType: 'file',
      errors: undefined
    }
  },
  '_root$Ref': $Ref {
    path: '/private/tmp/',
    value: { definitions: [Object] },
    '$refs': [Circular *1],
    pathType: 'file',
    errors: undefined
  }
}
```

@P0lip that circular reference is something we fixed in our fork right? Can we get that back upstream somehow?

**Comment by philsturgeon** [comment]
Loads of circular issues fixed in 9.0.x so let's assume we're good here. If not, please reopen.
